src/clarity-angular/datagrid/providers/hideable-column.service.ts

```diff
@@ -0,0 +1,192 @@
+     *
+     * @returns {boolean}
+     */
+    public get canHideNextColumn(): boolean {
+        let hiddenColumns = this._columnList.filter(col => col !== undefined).filter(col => col.hidden);
+        return ( this._columnList.length - hiddenColumns.length > 1 );
+    }
+
+    /**********
+     *
+     * @property checkForAllColumnsVisible
+     *
+     * @description
+     * For when you need to know if the datagrid's columns are all showing.
+     *
+     * @return {boolean}
+     */
+    public get checkForAllColumnsVisible(): boolean {
+        let allColumns = this._columnList.filter(col => col);
+        let hiddenColumns = this._columnList.filter(col => {
+            if ( col && !col.hidden ) {
+                return col;
+            }
+        });
+        return ( allColumns.length === hiddenColumns.length );
+    }
+
+    /***********
+     * @property columnListChange
+     *
+     * @description
+     * A public property that enables subscribers to hear updates to the column map.
+     * Use this if you need to do something whenever the Datagrid's column list is changed (i.e *ngIf on a column).
+     *
+     * @returns {Observable<DatagridHideableColumn[]>}
+     */
+    public get columnListChange(): Observable<DatagridHideableColumn[]> {
+        return this._columnListChange.asObservable();
+
+    }
+
+    /**********
+     *
+     * @function getColumns
+     *
+     * @description
+     * Public function that returns the current list of columns. I needed an array of to iterate on in the RowRenderer
+     * but subscribing to the _columnListChange changes did not seem like the correct way to get it.
+     *
+     * @returns {DatagridColumn[]}
+     */
+    public getColumns(): DatagridHideableColumn[] {
+        return this._columnList;
+    }
+
+    /**********
+     * @function showHiddenColumns
+     *
+     * @description
+     * Iterate through the current _columnList:
+     * - if it has a DatagridHideableColumn and is hidden then show it.
+     * - if it's DatagridHideableColumn was previously the last column visible, turn that flag off.
+     *
+     */
+    public showHiddenColumns() {
+        this._columnList.forEach(( col ) => {
+            if ( col && col.hidden === true ) {
+                col.hidden = false;
+            }
+
+            if ( col && col.lastVisibleColumn ) {
+                col.lastVisibleColumn = false;
+            }
+        });
+    }
+
+    /**
+     * @function updateColumnList
+     *
+     * @param columns: DatagridColumn[]
+     *
+     * @description
+     * Creates an array of DatagridHideableColumn's || null based column array passed as param.
+     * Is dependent on the order in @ContentChildren in Datagrid.
+     *
+     * @param columns
+     *
+     */
+    public updateColumnList( columns: DatagridHideableColumn[] ) {
+        this._columnList.length = 0; // clear the list
+        columns.forEach(col => this._columnList.push(col)); // Update it with the new
```

Comment:
Why not just `this._columnList = columns`? Do you really need to manually clone the array?